### PR TITLE
Typescript Demo

### PR DESCRIPTION
**Pull request by jaredponn**
What was done (excerpt shortened from `./build.nix`)
- TS demo project based off of the PureScript / Rust project which uses cardano-serialization-lib and ogmios. Running the tests requires a JSON *RTS configuration file* -- see `./src/rtsconfig.ts` for details in the environment variable `TEST_RTS_CONFIG`. If this is not provided, the tests will automatically spin everything up for you.
  
- The shell script `demo-rts` can be used to start a test RTS (ogmios + plutip) for you, and it will also provide a RTS configuration file for running the tests.

### CODE REVIEW

File: pre-commit.nix [hunk 1]
```diff
@@ -25,6 +25,19 @@
...
+              # See NOTE(jaredponn) below
+              files = ''^.*\.(ts|tsx)$'';
+            };
+            denofmt = {
+              enable = true;
+              # NOTE(jaredponn): We follow the default files this formats, except
+              # we exclude markdown, js, and json because those would mess with
+              # already existing things.
+              #   [1] https://docs.deno.com/runtime/manual/tools/formatter
+              files = ''^.*\.(ts|tsx)$'';
```
Comment (bladyjoker) [inline pre-commit.nix:39]: wth is this deno, sounds like it's trying to lint and format EVERYTHING

File: transactions/demo-typescript/build.nix [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      tsFlake = inputs.flake-lang.lib."${system}".typescriptFlake
+        rec
+        {
+          name = "demo";
+          src = ./.;
+
+          testTools = [
+            inputs'.plutip.packages."plutip-core:exe:local-cluster"
+            inputs'.ogmios.packages."ogmios:exe:ogmios"
+            demoRts
```
Comment (bladyjoker) [inline transactions/demo-typescript/build.nix:44]: I like!

File: transactions/demo-typescript/demo-rts.sh [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+WALLET_DIR="$(mktemp -d)"
+export NODE_WORKING_DIR
+NODE_WORKING_DIR="$(mktemp -d)"
+
+# The directory for plutip's local cluster information see [1]
+export LOCAL_CLUSTER_INFO_JSON="./local-cluster-info.json"
+
+# Ensure $LOCAL_CLUSTER_INFO_JSON exists because plutip doesn't like it if this
+# file doesn't already exist.
+touch "$LOCAL_CLUSTER_INFO_JSON"
```
Comment (bladyjoker) [inline transactions/demo-typescript/demo-rts.sh:47]: What? That doesn't sound right and I implemented the file output :D

File: transactions/demo-typescript/demo-rts.sh [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+# Start the local cluster in the background
+1>&2 echo "demo-rts: starting plutip (local-cluster) logging to ./local-cluster.log"
+local-cluster \
+    --num-wallets 1 \
```
Comment (bladyjoker) [inline transactions/demo-typescript/demo-rts.sh:52]: Can we make this into a parameter! Just thinking ahead a bit

File: transactions/demo-typescript/demo-rts.sh [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+1>&2 echo "demo-rts: starting ogmios in the background with host \`$OGMIOS_HOST\` on port \`$OGMIOS_PORT\` logging to ./ogmios.log"
+# shellcheck disable=SC2001
+ogmios \
+    --host "$OGMIOS_HOST" \
+    --port "$OGMIOS_PORT" \
+    --node-socket "$NODE_SOCKET" \
+    --node-config "$(echo "$NODE_SOCKET" | sed -e 's/\.socket$/\.config/')" \
+    2>&1 | tee ogmios.log 1>&2 \
+    &
```
Comment (bladyjoker) [inline transactions/demo-typescript/demo-rts.sh:102]: you might need to run a loop that checks whether Ogmios is synced

File: transactions/demo-typescript/src/lib/index.ts
Comment (bladyjoker) [inline transactions/demo-typescript/src/lib/index.ts:1]: Extract generic tx building stuff from the protocol specifics

File: transactions/demo-typescript/src/tests/demo-test.ts [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+  // NOTE(jaredponn): we use the same cluster for all tests
+  before(async () => {
+    if (testRtsConfigFilePath !== undefined) {
+      await initRtsConfig(testRtsConfigFilePath);
+    } else {
+      demoRts = spawn("demo-rts", [], {
+        stdio: ["ignore", "pipe", "ignore"],
+        detached: true,
+        env: Object.assign({ "ADA": "10" }, process.env),
+      });
```
Comment (bladyjoker) [inline transactions/demo-typescript/src/tests/demo-test.ts:62]: Ideally, I'd like for the RTS to managed (spawn, kill) from the outside, and not from the testsuite.

For example, if I point this testsuite to mainnet? You can't really spawn or kill it :D It's not urgent or really important for this PR, but generally that's where I want to go with this.

**Review by bladyjoker — CHANGES_REQUESTED** [review]
Great work! Left some commentie comments

File: transactions/demo-typescript/demo-rts.sh (hunk 3 above)
Comment (jaredponn) [reply transactions/demo-typescript/demo-rts.sh:47]: Yeah that doesn't seem right. It works even if the file doesn't already exist!

File: transactions/demo-typescript/demo-rts.sh (hunk 5 above)
Comment (jaredponn) [reply transactions/demo-typescript/demo-rts.sh:102]: Er I'm not sure how to do this -- mind elaborating a bit if it isn't too much trouble? I don't think ogmios has a notion of being synced -- we probably meant the underlying cardano node is synced?

But (as far as I am aware!) I don't think ogmios provides an interface to query if the node is "fully synced" or not.

File: transactions/demo-typescript/src/lib/index.ts
Comment (jaredponn) [reply transactions/demo-typescript/src/lib/index.ts:1]: Alright, I think I did this. The new code has:

- `./tx-builder.ts` which is a thin wrapper over Cardano serialization library
- `./query.ts` which isolates the functionality we need to query the cardano node

File: transactions/demo-typescript/demo-rts.sh (hunk 5 above)
Comment (bladyjoker) [reply transactions/demo-typescript/demo-rts.sh:99]: I guess smtn with a tip? Like ledger tip or network tip. Doesn't matter, I just remember using some API call to understand whether Ogmios is synced with the underlying cardano-node.

File: transactions/demo-typescript/src/lib/query.ts [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+  /**
+   * Queries the execution units of the transaction.
+   *
+   * @param excessBudgetFactor - what to multiply the budgets by after
+   * evaluating it. See the warning {@link https://ogmios.dev/mini-protocols/local-tx-submission/#evaluating-transactions}
+   */
+  evaluateTx(
+    transaction: csl.Transaction,
+    excessBudgetFactor?: number,
+  ): Promise<TxExUnits>;
```
Comment (bladyjoker) [inline transactions/demo-typescript/src/lib/query.ts:41]: You'd probably want to have a Submit interface or smtn

**Review by bladyjoker — APPROVED** [review]

File: transactions/demo-typescript/src/lib/query.ts (hunk 7 above)
Comment (jaredponn) [reply transactions/demo-typescript/src/lib/query.ts]: I see, I split `Query` into two interfaces:
- `Query` for `Query`ing things
- `Submit` for `Submit` transactions.
Note that Ogmios allows one to do both, and hence implements both interfaces

File: transactions/demo-typescript/demo-rts.sh (hunk 5 above)
Comment (jaredponn) [reply transactions/demo-typescript/demo-rts.sh:99]: Hm interesting.

Apparently the `/health` endpoint will return to you exactly the percent that it's synchronized? I think I'll use this

image

Alternatively, I reckon that computing the percent that it's synced manually via the ledger tip + network tip (not sure if this is how it's done) probably isn't that difficult either.

**Comment by jaredponn** [comment]
Okay! I think I addressed the issues! I'll merge this and as more problems show up, we can file more issues :)

**Pull request merged**